org.spoofax.jsglr2/src/main/java/org/spoofax/jsglr2/incremental/IncrementalReduceManager.java

```diff
@@ -0,0 +1,55 @@
+package org.spoofax.jsglr2.incremental;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.metaborg.parsetable.IParseTable;
+import org.metaborg.parsetable.actions.IReduce;
+import org.spoofax.jsglr2.incremental.parseforest.IncrementalParseForest;
+import org.spoofax.jsglr2.parseforest.IDerivation;
+import org.spoofax.jsglr2.parseforest.ParseForestConstruction;
+import org.spoofax.jsglr2.parseforest.ParseForestManager;
+import org.spoofax.jsglr2.parser.AbstractParse;
+import org.spoofax.jsglr2.reducing.ReduceManager;
+import org.spoofax.jsglr2.stack.AbstractStackManager;
+import org.spoofax.jsglr2.stack.IStackNode;
+import org.spoofax.jsglr2.stack.StackLink;
+import org.spoofax.jsglr2.stack.paths.StackPath;
+
+public class IncrementalReduceManager
+//@formatter:off
+   <ParseForest extends IncrementalParseForest,
+    ParseNode   extends ParseForest,
+    Derivation  extends IDerivation<ParseForest>,
+    StackNode   extends IStackNode,
+    Parse       extends AbstractParse<ParseForest, StackNode>>
+//@formatter:on
+    extends ReduceManager<ParseForest, ParseNode, Derivation, StackNode, Parse> {
+
+    public IncrementalReduceManager(IParseTable parseTable,
+        AbstractStackManager<ParseForest, StackNode, Parse> stackManager,
+        ParseForestManager<ParseForest, ParseNode, Derivation> parseForestManager,
+        ParseForestConstruction parseForestConstruction) {
+        super(parseTable, stackManager, parseForestManager, parseForestConstruction);
+    }
+
+    @Override protected void doReductionsHelper(Parse parse, StackNode stack, IReduce reduce,
+        StackLink<ParseForest, StackNode> throughLink) {
+
+        List<StackPath<ParseForest, StackNode>> paths = stackManager.findAllPathsOfLength(stack, reduce.arity());
+
+        if(throughLink != null)
+            paths = paths.stream().filter(path -> path.contains(throughLink)).collect(Collectors.toList());
+
+        if(paths.size() > 1)
+            // TODO this cast should not be necessary, but `Parse extends IncrementalParse<StackNode>` is not allowed
```

Comment:
As discussed, you might be able to prevent the cast here by introducing a `IIncrementalParse` interface. Then `IncrementalParse` would implement that and you could use the interface as type bound within this `ReduceManager`.

Reply:
Thanks again for the suggestion, it's done :slightly_smiling_face: While I was at it, I fixed a similar case in the `IncrementalParseForestManager`.